### PR TITLE
Gemma-7b&&Gemma-2b

### PR DESCRIPTION
**Pull request by intellinjun**
## Type of Change

feature or bug fix or documentation or others
API changed or not

## Description

detail description 
Issues: xxx

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Comment by intellinjun** [comment]
image
f32 inference result

**Comment by intellinjun** [comment]
image
gemma_2b inference result

**Comment by intellinjun** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


model_name | model_type | piqa
-- | -- | --
gemma_2b | weight_dtype f32 | 77.09
w4cint8g32 | 75.68
pytorch | 77.3
gemma_7b | weight_dtype f32 | 80.25
w4cint8g32 | 78.56
pytorch | 81.2



</body>

</html>

**Comment by intellinjun** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


model_name | model_type | piqa
-- | -- | --
gemma_2b | weight_dtype f32 | 77.09
gemma_2b | w4cint8g32 | 75.68
gemma_2b | pytorch | 77.3
gemma_7b | weight_dtype f32 | 80.25
gemma_7b | w4cint8g32 | 78.56
gemma_7b | pytorch | 81.2



</body>

</html>

**Comment by intellinjun** [comment]
> <style> </style>
> model_name	model_type	piqa
> gemma_2b	weight_dtype f32	77.09
> gemma_2b	w4cint8g32	75.68
> gemma_2b	pytorch	77.3
> gemma_7b	weight_dtype f32	80.25
> gemma_7b	w4cint8g32	78.56
> gemma_7b	pytorch	81.2

here is gemma7b&2b accuray,the inference result of gemma7b is bad when quantize config is weight_dtype int4 compute dtype int8 group_size 128, but gemma2b's result looks good.

File: neural_speed/convert/convert_bloom.py [hunk 1]
```diff
@@ -106,6 +106,7 @@ def main(args_in: Optional[List[str]] = None) -> None:
     fout.write(struct.pack("i", 0))
     fout.write(struct.pack("i", 0))  # n_experts
     fout.write(struct.pack("i", 0))  # n_expert_used
+    fout.write(struct.pack("i", 0)) # n_embd_head_k for gemma
```
Comment (airMeng) [inline neural_speed/convert/convert_bloom.py:109]: can we have some wiser solution instead of placeholder for each model?

Comment (intellinjun) [reply neural_speed/convert/convert_bloom.py:109]: Use gguf format is a solution, However, gguf is not the latest version of the pip package, and cannot support the latest model

File: tests/model-test/cpp_graph_inference.sh [hunk 2]
```diff
@@ -432,7 +439,7 @@ function main() {
                     else
                         real_ctx=$ctx # TODO(Zhenzhong): use same ctx for  chatglm & baichuan
                         [[ "${model}" == "chatglm2" || "${model}" == "chatglm-6b" ||
-                            "${model}" == "baichuan-13b" || "${model}" == "baichuan2-13b" ]] && real_ctx=1300
+                            "${model}" == "baichuan-13b" || "${model}" == "baichuan2-13b" ]] && real_ctx=2048
```
Comment (a32543254) [inline tests/model-test/cpp_graph_inference.sh:442]: why change baichuan real_ctx here ?

**Review by a32543254 — APPROVED** [review]
LGTM

File: tests/model-test/cpp_graph_inference.sh (hunk 2 above)
Comment (intellinjun) [reply tests/model-test/cpp_graph_inference.sh:442]: Zhenzhong's pr can automatically allocate memory, so I want to try to use baichuan2-13b inference with 2012 inputs.

File: neural_speed/core/layers/ip_fusion_ffn.cpp [hunk 3]
```diff
@@ -684,6 +684,11 @@ bool bestla_fusion_FFN_SiLu_f32f32_support(void* w1ptr, void* w2ptr, void* w3ptr
   return ffn_3w::bestla_fusion_ffn_f32f32_support(w1ptr, w2ptr, w3ptr, seq, fin, fmid, fout);
 }
 
+bool bestla_fusion_FFN_Gelu_Mul_f32f32_support(void* w1ptr, void* w2ptr, void* w3ptr, int seq, int fin, int fmid,
```
Comment (zhentaoyu) [inline neural_speed/core/layers/ip_fusion_ffn.cpp:687]: Does it mean gemma FFN-8W matmul? or I misunderstanding? And what is the difference between `Gelu_Mul` and `Gelu`?

**Review by zhentaoyu — APPROVED** [review]

File: docs/supported_models.md [hunk 4]
```diff
@@ -271,6 +271,17 @@ Neural Speed supports the following models:
     <td> </td>
     <td> </td>
     <td>Latest</td>
+  </tr>
+    <tr>
+    <td><a href="https://huggingface.co/google/gemma-2b-it" target="_blank" rel="noopener noreferrer">gemma-2b-it </a>,
+    <a href="https://huggingface.co/google/gemma-7b" target="_blank" rel="noopener noreferrer">gemma-7b</a></td>
+    <td>✅</td>
```
Comment (a32543254) [inline docs/supported_models.md:278]: format is worry, please fix the md format.

File: neural_speed/application/CMakeLists.txt [hunk 5]
```diff
@@ -97,6 +98,9 @@ set(mymap_phi 16)
 set(mymap_stablelm 17)
 set(mymap_whisper 18)
 set(mymap_mixtral 19)
+set(mymap_gemma 20)
+
+
```
Comment (a32543254) [inline neural_speed/application/CMakeLists.txt:103]: remove unused line

File: docs/supported_models.md [hunk 6]
```diff
@@ -271,6 +271,17 @@ Neural Speed supports the following models:
...
+    <td> </td>
+    <td> </td>
+    <td>✅</td>
```
Comment (a32543254) [inline docs/supported_models.md:281]: and do we support gguf of this model ? 
https://huggingface.co/lmstudio-ai/gemma-2b-it-GGUF

File: docs/supported_models.md [hunk 7]
```diff
@@ -271,6 +271,17 @@ Neural Speed supports the following models:
     <td> </td>
     <td> </td>
     <td>Latest</td>
+  </tr>
+    <tr>
+    <td><a href="https://huggingface.co/google/gemma-2b-it" target="_blank" rel="noopener noreferrer">gemma-2b-it </a>,
```
Comment (a32543254) [inline docs/supported_models.md:276]: there is no structure difference between gemma-2b and gemma-2b-it  right ? 
[gemma-2b](https://huggingface.co/google/gemma-2b)

File: neural_speed/core/layers/ip_fusion_ffn.cpp (hunk 3 above)
Comment (intellinjun) [reply neural_speed/core/layers/ip_fusion_ffn.cpp:687]: FFN_Gelu_Mul is self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x)),and FFN_Gelu is  self.fc_out(self.act(self.fc_in(x)))

File: docs/supported_models.md (hunk 7 above)
Comment (intellinjun) [reply docs/supported_models.md:276]: yes, gemma-2b-it is fine tune from gemma-2b, they have no difference.

File: neural_speed/core/layers/ip_fusion_ffn.cpp (hunk 3 above)
Comment (zhentaoyu) [reply neural_speed/core/layers/ip_fusion_ffn.cpp:687]: my bad, `3W` here means three weights ptr rather than three times dimension...

**Review by Zhenzhong1 — APPROVED** [review]

File: docs/supported_models.md (hunk 6 above)
Comment (intellinjun) [reply docs/supported_models.md:285]: support gguf model convert from llama.cpp

File: neural_speed/application/CMakeLists.txt (hunk 5 above)
Comment (intellinjun) [reply neural_speed/application/CMakeLists.txt:103]: done

**Comment by intellinjun** [comment]
https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/103/artifact/report.html

gemma-2b performance, looks so bad.
@luoyu-intel , I profiled the gemma-2b performance, and find the main cost is on the innerproduct,  gemma-2b has two innerproduct, lm_head.weight[2048, 256000] and token_embd.weight[256000,2048], can you have solution to fix it?
image

**Comment by intellinjun** [comment]
new gemma2 performance, @luoyu-intel ,thanks.
https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/107/artifact/report.html

**Pull request merged**